aptos-move/aptos-vm/src/aptos_vm.rs

```diff
@@ -1628,12 +1628,24 @@ impl AptosVM {
             ChangeSetConfigs::unlimited_at_gas_feature_version(self.gas_feature_version);
 
         match write_set_payload {
-            WriteSetPayload::Direct(change_set) => VMChangeSet::try_from_storage_change_set(
-                change_set.clone(),
-                &change_set_configs,
-                resolver.is_delayed_field_optimization_capable(),
-            )
-            .map_err(|e| e.into_vm_status()),
+            WriteSetPayload::Direct(change_set) => {
+                // this transaction is never delayed field capable.
+                // it is required it requires restarting execution afterwards,
+                // which allows it to be used as last transaction in delayed_field_enabled context.
+                let change = VMChangeSet::try_from_storage_change_set_with_delayed_field_optimization_disabled(
+                    change_set.clone(),
+                    &change_set_configs,
+                )
+                .map_err(|e| e.into_vm_status())?;
+
+                // validate_waypoint_change_set checks that this is true, so we only log here.
+                if !Self::should_restart_execution(&change) {
+                    error!(
+                        "[aptos_vm] direct write set in delayed_field_optimization_capable context finished without requiring should_restart_execution");
```

Comment:
why in delayed_field_optimization_capable context? capable means that the code to handle is there? (because control flow arrives here as long as the transaction does not emit epoch change event, right?)

Reply:
added explanation